spring-pulsar/src/main/java/org/springframework/pulsar/core/PulsarTemplate.java

```diff
@@ -115,22 +176,49 @@ private CompletableFuture<MessageId> doSendAsync(String topic, T message,
 			ProducerBuilderCustomizer<T> producerCustomizer) throws PulsarClientException {
 		final String topicName = ProducerUtils.resolveTopicName(topic, this.producerFactory);
 		this.logger.trace(() -> String.format("Sending msg to '%s' topic", topicName));
-		final Producer<T> producer = prepareProducerForSend(topic, message, messageRouter, producerCustomizer);
-		TypedMessageBuilder<T> messageBuilder = producer.newMessage().value(message);
-		if (typedMessageBuilderCustomizer != null) {
-			typedMessageBuilderCustomizer.customize(messageBuilder);
-		}
-		return messageBuilder.sendAsync().whenComplete((msgId, ex) -> {
-			if (ex == null) {
-				this.logger.trace(() -> String.format("Sent msg to '%s' topic", topicName));
-				// TODO success metrics
-			}
-			else {
-				this.logger.error(ex, () -> String.format("Failed to send msg to '%s' topic", topicName));
-				// TODO fail metrics
+
+		PulsarMessageSenderContext senderContext = PulsarMessageSenderContext.newContext(topicName, this.beanName);
+		Observation observation = newObservation(senderContext);
+		try {
+			observation.start();
```

Comment:
I would like to clean this up a bit (here and in the listener counterpart) so that the try/observe/catch is around a singe method that "does the thing". As it is, the really decreases readability.

Reply:
Will revisit in #147